### PR TITLE
Creating TLRU Cache TimedCacheDataSource

### PR DESCRIPTION
**Pull request by vilanovi**
New DataSource for a fast setup of a TLRU Cache (Time aware Least Recently Used).

- This is mostly a wrapper for an existing DataSource
- It uses the KeyQuery protocol to get a key for each query to cache stuff
- If not passed a KeyQuery, it just forwards everything to the original data source

This is a nice cache layer to be used on top of network calls, or complex database-based data sources, to speed up things.

Once approved by the team, we can then document it in the mjcore-documentation repo and implement it in android as well.

### CODE REVIEW

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    }
+    
+    @discardableResult
+    public func put(_ value: T? = nil, in query: Query) -> Future<T> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.put(value, in: query).filter { object in
+                    self.objects[query.key] = (object, Date())
+                }.recover { error in
+                    return Future(error)
```
Comment (vilanovi) [inline MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: I'm not sure what should we do in this case:

a) Should we remove any previously cached values for that key here?
b) Should we ignore it and do nothing?

In any case, the code as it is now it is doing nothing and we should remove the "recover" block unless we want to apply the option a).

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    }
+    
+    @discardableResult
+    public func putAll(_ array: [T], in query: Query) -> Future<[T]> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.putAll(array, in: query).filter { array in
+                self.arrays[query.key] = (array, Date())
+                }.recover { error in
+                    return Future(error)
```
Comment (vilanovi) [inline MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: I'm not sure what should we do in this case:

a) Should we remove any previously cached values for that key here?
b) Should we ignore it and do nothing?

In any case, the code as it is now it is doing nothing and we should remove the "recover" block unless we want to apply the option a).

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+    
+    @discardableResult
+    public func delete(_ query: Query) -> Future<Void> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.delete(query).filter { _ in
+                self.objects[query.key] = nil
+            }
```
Comment (vilanovi) [inline MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:121]: What if error? should we delete the current cached object anyway?

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+    
+    @discardableResult
+    public func deleteAll(_ query: Query) -> Future<Void> {
+        switch query {
+        case let query as KeyQuery:
+            return dataSource.delete(query).filter { _ in
+                self.arrays[query.key] = nil
+            }
```
Comment (vilanovi) [inline MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:134]: What if error? should we delete the current cached object anyway?

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 1 above)
Comment (vilanovi) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: Btw, I've seen is way better to use the `then` and `fail` future methods than `filter` and `recover`. Applying this upgrade to everywhere in the code.

Comment (joselufo) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: We should remove any previously cached values as we are not aware the business logic of the child datasource.

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 2 above)
Comment (joselufo) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: We should remove any previously cached values as we are not aware the business logic of the child datasource.

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 3 above)
Comment (joselufo) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:121]: We should delete cached values.

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 4 above)
Comment (joselufo) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:134]: We should delete cached values.

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+import Foundation
+
+fileprivate func print(_ object: String) {
```
Comment (joselufo) [inline MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:19]: I don't see value to have this code here. 

@newboadki is working in a logger protocol. We should remove this code and when @newboadki finish the logger we can update it

**Review by joselufo — CHANGES_REQUESTED** [review]

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 1 above)
Comment (vilanovi) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: makes sense. Will update it as you say. If error on any get/put/delete, any cached object will be removed.

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 2 above)
Comment (vilanovi) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift]: will remove cached object on any error get/put/Delete.

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 5 above)
Comment (vilanovi) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:19]: goal was to let the user know that values weren't cached as queries don't conform to `KeyQuery`. However, these logs are not valuable on a production app.

**Comment by vilanovi** [comment]
Ok, changes applied.

**Review by RubenVot — APPROVED** [review]

File: MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift (hunk 5 above)
Comment (joselufo) [reply MJSwiftCore/Classes/Repository/TimedCacheDataSource.swift:19]: Ok.

**Review by joselufo — APPROVED** [review]
All OK for me.

**Comment by vilanovi** [comment]
reference updated:

https://github.com/mobilejazz/mjcore-reference/blob/master/TimedCacheDataSource.md

**Pull request merged**